core/state/database.go

```diff
@@ -250,6 +252,118 @@ func (tds *TrieDbState) Copy() *TrieDbState {
 	return &cpy
 }
 
+// PutTombstoneForDeletedAccount - placing tombstone only if given account has storage in database
+func PutTombstoneForDeletedAccount(db ethdb.MinDatabase, addrHash []byte) error {
+	if len(addrHash) != common.HashLength {
+		return nil
+	}
+
+	var boltDb *bolt.DB
+	if hasKV, ok := db.(ethdb.HasKV); ok {
+		boltDb = hasKV.KV()
+	} else {
+		return fmt.Errorf("only Bolt supported yet, given: %T", db)
+	}
+
+	hasIH := false
+	if err := boltDb.View(func(tx *bolt.Tx) error {
+		k, _ := tx.Bucket(dbutils.IntermediateTrieHashBucket).Cursor().Seek(addrHash)
+		if !bytes.HasPrefix(k, addrHash) {
+			k = nil
+		}
+		if k != nil {
+			hasIH = true
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if !hasIH {
+		return nil
+	}
+
+	return db.Put(dbutils.IntermediateTrieHashBucket, common.CopyBytes(addrHash), []byte{})
+}
+
+func ClearTombstonesForNewStorage(db ethdb.MinDatabase, storageKeyNoInc []byte) error {
+	var boltDb *bolt.DB
+	if hasKV, ok := db.(ethdb.HasKV); ok {
+		boltDb = hasKV.KV()
+	} else {
+		return fmt.Errorf("only Bolt supported yet, given: %T", db)
+	}
+
+	var toPut [][]byte
+	toDelete := map[string]struct{}{}
+
+	if err := boltDb.View(func(tx *bolt.Tx) error {
+		c := tx.Bucket(dbutils.IntermediateTrieHashBucket).Cursor()
+
+		i := common.HashLength
+		var k, v []byte
+		for ; i < len(storageKeyNoInc); i++ {
+			k, v = c.Seek(storageKeyNoInc[:i])
+			if k == nil {
+				return nil
+			}
+
+			isTombstone := v != nil && len(v) == 0
+			if !isTombstone {
+				continue
+			}
+
+			if !bytes.HasPrefix(storageKeyNoInc, k) {
+				continue
+			}
+
+			break
+		}
+
+		for ; k != nil; k, v = c.Next() {
+			if !bytes.HasPrefix(k, storageKeyNoInc[:i]) {
+				k = nil
+			}
+			if k == nil {
```

Comment:
Since `k` only changes in the `for` clause, and `k != nil` is check, you don't need this extra check, if you just break in the `if` above